morebits.js

```diff
@@ -2930,36 +2930,72 @@ Morebits.wiki.page = function(pageName, status) {
+	 * watch unless `watchlistExpiry` is provided.
+	 * - `false`|`'no'`|`'nochange'`: watchlist status of the page (including expiry) will not be changed.
+	 * - `'default'`|`'preferences'`: watchlist status of the page will be
+	 * set based on the user's preference settings when the action is
+	 * called. Defaults to an indefinite watch unless `watchlistExpiry` is
+	 * provided.
+	 * - `'unwatch'`: explicitly unwatch the page.
+	 * - Any other `string` or `number`: watch page until the specified
+	 * time, deferring to `watchlistExpiry` if provided.
+	 * @param {string|number} [watchlistExpiry=infinity] - A date-like string or number.
+	 * Can be relative (2 weeks) or other similarly date-like (i.e. NOT "potato"):
+	 * ISO 8601: 2038-01-09T03:14:07Z
+	 * MediaWiki: 20380109031407
+	 * UNIX: 2147483647
+	 * SQL: 2038-01-09 03:14:07
+	 * Can also be `infinity` or infinity-like (`infinite`, `indefinite`, and `never`).
+	 * See {@link https://phabricator.wikimedia.org/source/mediawiki-libs-Timestamp/browse/master/src/ConvertibleTimestamp.php;4e53b859a9580c55958078f46dd4f3a44d0fcaa0$57-109?as=source&blame=off}
 	 */
-	this.setWatchlist = function(watchlistOption) {
-		if (!watchlistOption || watchlistOption === 'no' || watchlistOption === 'nochange') {
-			ctx.watchlistOption = 'nochange';
-		} else if (watchlistOption === 'default' || watchlistOption === 'preferences') {
-			ctx.watchlistOption = 'preferences';
-		} else if (watchlistOption === 'unwatch') {
-			ctx.watchlistOption = 'unwatch';
-		} else {
-			ctx.watchlistOption = 'watch';
-			if (typeof watchlistOption === 'number' || (typeof watchlistOption === 'string' && watchlistOption !== 'yes')) {
-				ctx.watchlistExpiry = watchlistOption;
-			}
+	this.setWatchlist = function(watchlistOption, watchlistExpiry) {
+		watchlistExpiry = typeof watchlistExpiry !== 'undefined' ? watchlistExpiry : 'infinity';
+		switch (watchlistOption) {
+			case 'nochange':
+			case 'no':
+			case false:
+			case undefined:
+				ctx.watchlistOption = 'nochange';
+				// The MW API allows for changing expiry with nochange (as "nochange" refers to the binary status),
+				// but by keeping this null it will default to any existing expiry, ensure there is actually "no change."
+				ctx.watchlistExpiry = null;
+				break;
+			case 'unwatch':
+				// expiry unimportant
+				ctx.watchlistOption = 'unwatch';
+				break;
+			case 'preferences':
+			case 'default':
+				ctx.watchlistOption = 'preferences';
+				// The API allows an expiry here, but there is as of yet (T265716)
+				// no expiry preference option, so it's a bit devoid of context.
+				ctx.watchlistExpiry = watchlistExpiry;
+				break;
+			case 'watch':
+			case 'yes':
+			case true:
+				ctx.watchlistOption = 'watch';
+				ctx.watchlistExpiry = watchlistExpiry;
+				break;
+			default: // Not really a "default" per se but catches "any other string"
+				ctx.watchlistOption = 'watch';
+				ctx.watchlistExpiry = watchlistExpiry || watchlistOption;
```

Comment:
`watchlistExpiry || watchlistOption`  ?

Reply:
Mmm, yeah, good call.  This was a holdover from a previous structuring, but I was trying to be nice by defaulting to the explicit `watchlistExpiry` if both were provided.  With the overwriting to `infinity`, though, this clearly stopped working!  I've removed it and just used `watchlistOption`.

I suppose another idea (more in line with what I had initially) would be to remove the `typeof watchlistExpiry === 'undefined'` line and use that in just the `watch` and `preferences` options, then this could be a `typeof` ternary.  Probably excessive.